### PR TITLE
Document locking and unlocking swarm managers

### PR DESCRIPTION
**Pull request by mstanleyjones**
### Describe the proposed changes

Document locking a swarm.

### Unreleased project version

Engine 1.13

### Related issue

#689 

### Related issue or PR in another project

https://github.com/docker/docker/pull/27967/

### Please take a look

@cyli @aaronlehmann @diogomonica @rahim888

### CODE REVIEW

**Comment by johndmulhausen** [comment]
Deploy preview failed.

Built with commit 96672c0fe87de44bdaac9587eea41df4c52fcc90

https://app.netlify.com/sites/docker-docs/deploys/5834aace6686742e1efa016b

File: engine/swarm/swarm_manager_locking.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+When Docker restarts, the encryption key needed to encrypt and decrypt RAFT logs
+is loaded into each manager node's running memory from disk. Docker 1.13
+introduces the ability to protect the encryption key at rest by _autolocking_
+your swarm.
+
+When Docker restarts or a new manager attempts to join the swarm, or whenever
+the encryption key needs to be loaded from disk to a node's memory, you must
+[unlock the swarm](swarm_manager_locking.md#unlock-a-swarm) first, using a key
+provided by you or generated by Docker when the swarm was locked. You can rotate
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: Users cannot provide your own key - it's similar to join tokens.  It is generated by docker when you enable auto-locking, or when you rotate the unlock key.

File: engine/swarm/swarm_manager_locking.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+In Docker 1.13 and higher, the RAFT logs used by swarm managers are encrypted on
+disk by default. This at-rest encryption protects your service's configuration
+and data from attackers who gain access to the encrypted RAFT logs. One of the
+reasons this feature was introduced was in support of the new [Docker
+secrets](secrets.md) feature.
+
+When Docker restarts, the encryption key needed to encrypt and decrypt RAFT logs
+is loaded into each manager node's running memory from disk. Docker 1.13
+introduces the ability to protect the encryption key at rest by _autolocking_
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: It also protects the TLS key (the same unlock key is used to encrypt the TLS key) that the manager node uses to connect to the rest of the swarm, because having a mTLS connection to the rest of the swarm means that you can get all the swarm data (including secrets).

File: engine/swarm/swarm_manager_locking.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+your swarm.
+
+When Docker restarts or a new manager attempts to join the swarm, or whenever
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: When a new manager joins, the unlock key does not need to be provided, because that new manager does not have any data on disk that needs to be loaded.  The new manager will obtain the the unlock key from the raft leader via propagation of the rest of the raft data.

File: engine/swarm/swarm_manager_locking.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+command and provide the following key:
+
+    SWMKEY-1-+MrE8NgAyKj5r3NcR4FiQMdgu+7W72urH0EZeSmP/0Y
+
+Please remember to store this key in a password manager, since without it you
+will not be able to restart the manager.
+```
+
+To disable autolock, set `--autolock` to `false`. The encryption key used to
+read and write RAFT logs will be stored unencrypted on disk, so this
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: Maybe "The mutual TLS key and the encryption key used to ..."?

**Comment by cyli** [comment]
Thanks for the great work and quick turnaround on this @mstanleyjones!  A couple of nitpicks, but this looks fantastic.

**Comment by mstanleyjones** [comment]
Addressed feedback so far.

File: engine/swarm/swarm_manager_locking.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+In Docker 1.13 and higher, the RAFT logs used by swarm managers are encrypted on
+disk by default. This at-rest encryption protects your service's configuration
+and data from attackers who gain access to the encrypted RAFT logs. One of the
+reasons this feature was introduced was in support of the new [Docker
+secrets](secrets.md) feature.
+
+When Docker restarts, both the TLS key used to encrypt communication among
+swarm nodes, and the key used to encrypt and decrypt RAFT logs on disk, are
+loaded into each manager node's running memory from disk. Docker 1.13 introduces
+the ability to protect the encryption key at rest by _autolocking_ your swarm.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: `by allowing you to take ownership of this key and require manual unlocking of your managers. We call this feature autolocking`.

File: engine/swarm/swarm_manager_locking.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+When Docker restarts or whenever the encryption key needs to be loaded from disk
+to a node's memory, you must
+[unlock the swarm](swarm_manager_locking.md#unlock-a-swarm) first, using a key
+generated by Docker when the swarm was locked. When a new node joins the swarm,
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: I don't think this `When a new node joins the swarm...` part.

File: engine/swarm/swarm_manager_locking.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+    SWMKEY-1-+MrE8NgAyKj5r3NcR4FiQMdgu+7W72urH0EZeSmP/0Y
+
+Please remember to store this key in a password manager, since without it you
+will not be able to restart the manager.
+```
+
+To disable autolock, set `--autolock` to `false`. The mutual TLS key and the
+encryption key used to read and write RAFT logs will be stored unencrypted on
+disk, so this configuration is not recommended.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: `so, depending on the level of security you're striving for, this configuration might not be recommended.`

File: engine/swarm/swarm_manager_locking.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+To unlock a locked swarm, use `docker swarm unlock`.
+
+```bash
+$ docker swarm unlock
+
+Please enter unlock key:
+```
+
+Enter the encryption key that was emitted when you locked the swarm, and the
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: emitted -> generated.

File: engine/swarm/swarm_manager_locking.md (hunk 6 above)
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: I think making it clear here that the unlock key does not need to be provided when adding a manager (only the join token is needed, because the unlock key is propagated via the raft) could be useful - the user might think it needs to be added otherwise, if they're not sure when loading from disk is necessary.

Nitpick though, "the is" is missing a word :)  Maybe "the key is propagated..."?

File: engine/swarm/swarm_manager_locking.md (hunk 5 above)
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: @diogomonica  ownership of both the encryption and TLS key you mean?  Or of the unlock key?

File: engine/swarm/swarm_manager_locking.md (hunk 8 above)
Comment (mstanleyjones) [inline engine/swarm/swarm_manager_locking.md]: Well, it was generated but it was also emitted to standard output... let me see what I can do.

**Comment by mstanleyjones** [comment]
PTAL again. Thanks!

File: engine/swarm/swarm_manager_locking.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+In Docker 1.13 and higher, the RAFT logs used by swarm managers are encrypted on
+disk by default. This at-rest encryption protects your service's configuration
+and data from attackers who gain access to the encrypted RAFT logs. One of the
+reasons this feature was introduced was in support of the new [Docker
+secrets](secrets.md) feature.
+
+When Docker restarts, both the TLS key used to encrypt communication among swarm
+nodes, and the key used to encrypt and decrypt RAFT logs on disk, are loaded
+into each manager node's running memory from disk. Docker 1.13 introduces the
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: drop "from disk"?

File: engine/swarm/swarm_manager_locking.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+ability to protect the mutual TLS encryption key and the key used to encrypt and
+decrypt RAFT logs at rest (on disk), by allowing you to take ownership of this
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: drop "on disk"

File: engine/swarm/swarm_manager_locking.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+key and to require manual unlocking of your managers. This feature is called
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: This feature is called `autolock`.

File: engine/swarm/swarm_manager_locking.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+_autolocking_ your swarm.
+
+When Docker restarts or whenever the encryption key needs to be loaded from disk
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: "When a Docker restarts and needs to access the encrypted RAFT state on disk, you must..."

There aren't other motives to unlock a manager other than starting it (or restarting it).

File: engine/swarm/swarm_manager_locking.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+to a node's memory, you must
+[unlock the swarm](swarm_manager_locking.md#unlock-a-swarm) first, using a key
+generated by Docker when the swarm was locked. You can rotate this key at any time.
+
+>**Note**: You don't need to unlock the swarm when a new node joins the swarm,
+because the key is propagated to it over mutual TLS.. You only need to provide
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: two ".."

File: engine/swarm/swarm_manager_locking.md [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+the join token. The unlock key will be propagated as part of the RAFT log, which
+is encrypted at rest and transmitted over a mutual TLS connection.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: Do we need the second part? Maybe just this is enough:

```
**Note**: You don't need to unlock the swarm when a new node joins the swarm,
 +because the key is propagated to it over mutual TLS.
```

File: engine/swarm/swarm_manager_locking.md [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+## Initialize a swarm with autolocking enabled
+
+When you initialize a new swarm, you you can use the `--autolock` flag to
+automatically lock swarm manager nodes.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: `automatically enable autolocking of swarm manager nodes`. It doesn't actually "lock" anything.

File: engine/swarm/swarm_manager_locking.md [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.
+
+To unlock a swarm manager after it restarts, run the `docker swarm unlock`
+command and provide the following key:
+
+    SWMKEY-1-WuYH/IX284+lRcXuoVf38viIDK3HJEKY13MIHX+tTt8
+```
+
+Record the key in a safe place.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: Record -> Store?

Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: Also add: We recommend a password manager?

File: engine/swarm/swarm_manager_locking.md [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+When Docker restarts, a new manager node attempts to join the swarm, or Docker
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: please remove: `a new manager node attempts to join the swarm`

File: engine/swarm/swarm_manager_locking.md [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+Please remember to store this key in a password manager, since without it you
+will not be able to restart the manager.
+```
+
+To disable autolock, set `--autolock` to `false`. The mutual TLS key and the
+encryption key used to read and write RAFT logs will be stored unencrypted on
+disk. There is a trade-off between the risk of storing the encryption key
+unencrypted at rest and the convenience of being able to restart a swarm without
+needing to unlock the swarm.
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: `without needing to unlock each manager.`

There is no swarm-wide unlock. This is individual per manager (if all nodes go down, and you have a 3 manager swarm, you need to unlock each manager individually).

File: engine/swarm/swarm_manager_locking.md [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+```
+
+Enter the encryption key that was generated and shown in the command output when
+you locked the swarm or rotated the key, and the swarm unlocks.
+
+## View the current unlock key for a running swarm
+
+Consider a situation where your swarm is
+running as expected, then manager node becomes unavailable. You troubleshoot the
+problem and bring the physical node back online, but you need to rejoin the
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: You're not rejoining the node to the swarm. I think we should only use "join" when it's the first time a node is actually joining the swarm. Maybe rephrase:

`You troubleshoot the problem and bring the physical node back online, but you need to provide the key necessary to read the encrypted credentials and RAFT logs on disk.`

File: engine/swarm/swarm_manager_locking.md [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+To unlock a swarm manager after it restarts, run the `docker swarm unlock`
+command and provide the following key:
+
+    SWMKEY-1-8jDgbUNlJtUe5P/lcr9IXGVxqZpZUXPzd+qzcGp4ZYA
+
+Please remember to store this key in a password manager, since without it you
+will not be able to restart the manager.
+```
+
+If you ever need to view the unlock key in this way,
```
Comment (diogomonica) [inline engine/swarm/swarm_manager_locking.md]: Doesn't make a lot of sense, since rotating a swarm also displays the new key. I think we should remove it, or at least replace it with a warning around: please don't share this key anywhere, and store it securely, etc`

File: engine/swarm/swarm_manager_locking.md (hunk 15 above)
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: Maybe just "enable autolocking of swarm manager nodes", since "autolocking" implies automatic?

**Comment by mstanleyjones** [comment]
Addressed @diogomonica's feedback. Please let me know if you see more areas for improvement.

File: engine/swarm/swarm_manager_locking.md [hunk 21]
```diff
@@ -0,0 +1,157 @@
+---
+description: Automatically lock Swarm managers to protect encryption keys
+keywords:
+- swarm, manager, lock, unlock, autolock, encryption
+title: Lock your swarm to protect its encryption key
+---
+
+In Docker 1.13 and higher, the RAFT logs used by swarm managers are encrypted on
```
Comment (aaronlehmann) [inline engine/swarm/swarm_manager_locking.md]: In other docs we seem to use the capitalization `Raft`. This is also what the paper calls it.

File: engine/swarm/swarm_manager_locking.md [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+Error response from daemon: Swarm is encrypted and needs to be unlocked before it can be used. Please use "docker swarm unlock" to unlock it.
+```
+
+## Enable or disable autolock on an existing swarm
+
+To enable autolock on an existing swarm, set the `autolock` flag to `true`.
+
+```bash
+$ docker swarm update --autolock="true"
```
Comment (aaronlehmann) [inline engine/swarm/swarm_manager_locking.md]: The quotes around `true` are not necessary.

File: engine/swarm/swarm_manager_locking.md [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+```
+
+To disable autolock, set `--autolock` to `false`. The mutual TLS key and the
+encryption key used to read and write RAFT logs will be stored unencrypted on
+disk. There is a trade-off between the risk of storing the encryption key
+unencrypted at rest and the convenience of being able to restart a swarm without
+needing to unlock each manager.
+
+```bash
+$ docker swarm update --autolock="false"
```
Comment (aaronlehmann) [inline engine/swarm/swarm_manager_locking.md]: The quotes around `false` are not necessary.

File: engine/swarm/swarm_manager_locking.md [hunk 24]
```diff
@@ -0,0 +1,157 @@
+---
+description: Automatically lock Swarm managers to protect encryption keys
+keywords:
+- swarm, manager, lock, unlock, autolock, encryption
+title: Lock your swarm to protect its encryption key
+---
+
+In Docker 1.13 and higher, the RAFT logs used by swarm managers are encrypted on
+disk by default. This at-rest encryption protects your service's configuration
```
Comment (aaronlehmann) [inline engine/swarm/swarm_manager_locking.md]: We always encrypt the logs, but may or may not store the key alongside them. `--autolock=true` sets it up not to store the key.

File: engine/swarm/swarm_manager_locking.md (hunk 22 above)
Comment (aaronlehmann) [inline engine/swarm/swarm_manager_locking.md]: It's technically valid, but it's very atypical to have double quotes there. You could also write it as `docker "swarm" "update" "--autolock"=true`, but people don't do that either. So it looks like there are quotation marks in an arbitrary place where they do nothing, and I think it would look strange to people who understand the shell's syntax rules.

Comment (mstanleyjones) [inline engine/swarm/swarm_manager_locking.md]: OK, removed the quotes.

**Comment by mstanleyjones** [comment]
OK, I think this is ready to merge, just need some +1s.

**Comment by diogomonica** [comment]
LGTM

File: engine/swarm/swarm_manager_locking.md [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+and data from attackers who gain access to the encrypted Raft logs. One of the
+reasons this feature was introduced was in support of the new [Docker
+secrets](secrets.md) feature.
+
+When Docker restarts, both the TLS key used to encrypt communication among swarm
+nodes, and the key used to encrypt and decrypt Raft logs on disk, are loaded
+into each manager node's memory. Docker 1.13 introduces the ability to protect
+the mutual TLS encryption key and the key used to encrypt and decrypt Raft logs
+at rest, by allowing you to take ownership of this key and to require manual
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: Minor nitpick:  The first part of the sentence mentions 2 keys, and the second part says "take ownership of this key" - should it be "these keys"?  Or is this referring to the key-encrypting-key (then unlock key) (which hasn't been mentioned before).

File: engine/swarm/swarm_manager_locking.md [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+unlocking of your managers. This feature is called _autolock_.
+
+When Docker restarts, you must
+[unlock the swarm](swarm_manager_locking.md#unlock-a-swarm) first, using a key
```
Comment (cyli) [inline engine/swarm/swarm_manager_locking.md]: Non-blocking:  I don't know whether it'd be too much detail to mention that this is a key encrypting key - there are two other keys already mentioned in this section, and it might be useful to provide more disambiguation.  Not sure whether it'd just be more confusing, though.

**Comment by cyli** [comment]
One minor nitpick, but this LGTM.  Thanks @mstanleyjones!

**Comment by mstanleyjones** [comment]
Addressed @cyli feedback. If this is good to go, I will merge it after the rebase of `vnext-engine` branch is complete.
